622 tests depending on datasets are skipped

alberto-scolari opened the pull request:
This PR restructures the internal CI adding several features:
- testing inside a custom Docker container, with pre-compiled LPF and datasets (i.e., easily reproducible builds)
- hence, enabling tests with LPF and datasets
- tests against multiple GCC versions
- task-based description of CI jobs, with full parallelism and minimal dependencies (explicitly written)

@anyzelman This PR should have **high priority**, since a couple of incoming PRs depend on it. At the same time, existing PRs do **not** depend on it.

anyzelman commented:
By your comment, I assume ok to merge GMRES first? Is Benjamin's dependent on this one? Many thanks!

anyzelman commented:
> By your comment, I assume ok to merge GMRES first? Is Benjamin's dependent on this one? Many thanks!

@alberto-scolari ?

alberto-scolari commented:
@anyzelman it is ok to merge GMRES first, there's no interaction with this one and none also with Benjamin's

however, Benjamin's PR overlaps with this one, which is why we asked to merge this first

alberto-scolari commented:
> Two questions, @alberto-scolari @byjtew :
> 
> 1. is it correct that with the changes here, our gitlab-ci.yml becomes site-specific, and
What does "site-specific" mean?
If it means related to our internal infrastructure, than the answer is no, anyone running GitLab can run it; the only "site-specific" thing is indeed the Docker image we use, stored internally.

> 2. is it also correct that by publishing the image online we would allow external folks to run the gitlab CI (i.e., it would no longer be site-specific)?
I am not sure we can directly publish the image because of the datasets, for which the licenses must be accepted (as in the related ALP script). But we can publish a Dockerfile that packs everything; in such a case, the user has to accept the licenses and download the datasets (clearly we can automate this during the docker build). In summary: we cannot provide the binary image directly, but a straightforward way to build it.

anyzelman commented:
> > Two questions, @alberto-scolari @byjtew :
> > 
> > 1. is it correct that with the changes here, our gitlab-ci.yml becomes site-specific, and
> 
> What does "site-specific" mean? If it means related to our internal infrastructure, than the answer is no, anyone running GitLab can run it; the only "site-specific" thing is indeed the Docker image we use, stored internally.
> 
> > 2. is it also correct that by publishing the image online we would allow external folks to run the gitlab CI (i.e., it would no longer be site-specific)?
> 
> I am not sure we can directly publish the image because of the datasets, for which the licenses must be accepted (as in the related ALP script). But we can publish a Dockerfile that packs everything; in such a case, the user has to accept the licenses and download the datasets (clearly we can automate this during the docker build). In summary: we cannot provide the binary image directly, but a straightforward way to build it.

That's what I meant yes, thanks=) Then as part of this MR we ideally publish the dockerfile somewhere and include instructions where to get that and build the image that the CI script now requires? Let me know what if this requires web space somewhere (we have), if a free dockerfile registry does not suffice

alberto-scolari commented:
> That's what I meant yes, thanks=) Then as part of this MR we ideally publish the dockerfile somewhere and include instructions where to get that and build the image that the CI script now requires? Let me know what if this requires web space somewhere (we have), if a free dockerfile registry does not suffice

No registry needed nor web space, we can just create another repo ("ALP-CI" or smtg like that) with the Dockerfile (it's a Bash-like text file, currently 30 lines total), possibly the script to download the datasets and a README. And just point users to that in our docs here.

anyzelman commented:
> > That's what I meant yes, thanks=) Then as part of this MR we ideally publish the dockerfile somewhere and include instructions where to get that and build the image that the CI script now requires? Let me know what if this requires web space somewhere (we have), if a free dockerfile registry does not suffice
> 
> No registry needed nor web space, we can just create another repo ("ALP-CI" or smtg like that) with the Dockerfile (it's a Bash-like text file, currently 30 lines total), possibly the script to download the datasets and a README. And just point users to that in our docs here.

sounds great!

alberto-scolari commented on .gitlab-ci.yml:
```diff
@@ -1,29 +1,43 @@
+  image: ${CI_REGISTRY}/${CI_PROJECT_PATH}/lpf-ubuntu-20.04-gcc
+
+variables:
+  IS_MAIN_PIPELINE: "no"
```
The reason why I chose this pattern is because modifications to the rules are rare (I think they never happened since their introduction) and because at the same time there is only a few rule patterns that encompass multiple, related targets (e.g., build, install, unit, smoke, all of them for multiple configurations). In this context, I think it's cleaner and safer to "factor out" conditions into a variable that the jobs just read. If we make a change to the logic (e.g., trigger when the branch name also contains "release"), this likely involves entire groups of jobs, and can be applied only in one place.

alberto-scolari commented on tests/unit/parser.cpp:
```diff
@@ -197,15 +197,19 @@ bool readEdges(
 	// use utils parser
-	grb::utils::MatrixFileReader< void > citHepTh( "datasets/cit-HepTh.txt", false, true );
+	const char * const dataset_file = argv[ 1 ];
+	grb::utils::MatrixFileReader< void > citHepTh( dataset_file , false, true );
```
done it in the latest commit pushed on top

alberto-scolari commented on tests/unit/utilParserTest.cpp:
```diff
@@ -20,9 +20,12 @@
+		std::cout << "please, give path to west0497.mtx" << std::endl;
+		std::exit( 1 );
+	}
+	grb::utils::MatrixFileReader< double, unsigned short int > west( argv[ 1 ] );
```
done as well

alberto-scolari commented:
@anyzelman I implemented the suggested changes and rebased, if no more concern we can proceed merging

alberto-scolari commented:
> 



> > @anyzelman I implemented the suggested changes and rebased, if no more concern we can proceed merging
> 
> Hi @alberto-scolari -- the last point seems not yet addressed, maybe there's an unpushed commit somewhere?
> 
> (The last point being the documentation should point to https://github.com/Algebraic-Programming/ReproducibleBuilds to indicate how to get the image that would enable gitlab CI for external users/devs after this MR appears in develop)
> 
> Meanwhile finished the code review and pushed some minor fixes, welcome to check them as well

Your fixes are fine to me. I also pushed the updated documentation and also the Dockerfile's and the guide to

https://github.com/Algebraic-Programming/ReproducibleBuilds

so now we should have everything in place.

anyzelman commented:
Internal CI OK, code review OK, unit tests OK. Will merge
